core/src/main/java/org/elasticsearch/cluster/service/TaskBatching.java

```diff
@@ -0,0 +1,233 @@
+                synchronized (tasksPerExecutor) {
+                    LinkedHashSet<BatchingTask> existingTasks = tasksPerExecutor.get(batchingExecutor);
+                    if (existingTasks != null) {
+                        existingTasks.removeAll(toRemove);
+                        if (existingTasks.isEmpty()) {
+                            tasksPerExecutor.remove(batchingExecutor);
+                        }
+                    }
+                }
+                for (BatchingTask task : toRemove) {
+                    onTimeout(task, timeout);
+                }
+            }
+        });
+    }
+
+    /**
+     * Action to be implemented by the specific batching implementation
+     */
+    protected abstract void onTimeout(BatchingTask task, TimeValue timeout);
+
+    void runIfNotProcessed(BatchingTask updateTask) {
+        // if this task is already processed, the executor shouldn't execute other tasks (that arrived later),
+        // to give other executors a chance to execute their tasks.
+        if (updateTask.processed.get() == false) {
+            final List<BatchingTask> toExecute = new ArrayList<>();
+            final Map<String, List<BatchingTask>> processTasksBySource = new HashMap<>();
+            synchronized (tasksPerExecutor) {
+                LinkedHashSet<BatchingTask> pending = tasksPerExecutor.remove(updateTask.batchingKey);
+                if (pending != null) {
+                    for (BatchingTask task : pending) {
+                        if (task.processed.getAndSet(true) == false) {
+                            logger.trace("will process {}", task);
+                            toExecute.add(task);
+                            processTasksBySource.computeIfAbsent(task.source, s -> new ArrayList<>()).add(task);
+                        } else {
+                            logger.trace("skipping {}, already processed", task);
+                        }
+                    }
+                }
+            }
+
+            if (toExecute.isEmpty() == false) {
+                final String tasksSummary = processTasksBySource.entrySet().stream().map(entry -> {
+                    String tasks = updateTask.describeTasks(entry.getValue());
+                    return tasks.isEmpty() ? entry.getKey() : entry.getKey() + "[" + tasks + "]";
+                }).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
+
+                run(updateTask.batchingKey, toExecute, tasksSummary);
+            }
+        }
+    }
+
+    /**
+     * Action to be implemented by the specific batching implementation
+     */
+    protected abstract void run(Object batchingKey, List<? extends BatchingTask> tasks, String tasksSummary);
+
+    /**
+     * Represents a runnable task that supports batching.
+     * Implementors of TaskBatching can subclass this to add a payload to the task.
+     */
+    protected abstract class BatchingTask extends SourcePrioritizedRunnable {
```

Comment:
My preference goes to BatchedTask and have the taskIdentity become just task (next to the batching key). Obviously this falls deep into bikeshedding territory so feel totally free to reject and keep what you have.